rust/src/coordinator.rs

```diff
@@ -162,89 +120,107 @@ impl<N> Default for Coordinator<N> {
-        events, VecDeque<ProtocolEvent>;
-    );
-
+impl Coordinator {
     /// Create a coordinator. Fails if there is insufficient system entropy to generate secrets.
-    fn new() -> Result<Self, InitError>;
+    pub fn new() -> Result<Self, InitError> {
+        // crucial: init must be called before anything else in this module
+        sodiumoxide::init().or(Err(InitError))?;
+        let seed = randombytes(32);
+        Ok(Self {
+            seed,
+            ..Default::default()
+        })
+    }
+
+    /// Emit an event
+    pub fn emit_event(&mut self, event: ProtocolEvent) {
+        self.events.push_back(event);
+    }
 
-    /// Emit an event.
-    fn emit_event(&mut self, event: ProtocolEvent) {
-        self.events_mut().push_back(event);
+    /// Retrieve the next event
+    pub fn next_event(&mut self) -> Option<ProtocolEvent> {
+        self.events.pop_front()
     }
 
-    /// Retrieve the next event.
-    fn next_event(&mut self) -> Option<ProtocolEvent> {
-        self.events_mut().pop_front()
+    fn message_open(&self) -> MessageOpen<'_, '_> {
+        MessageOpen {
+            recipient_pk: &self.pk,
+            recipient_sk: &self.sk,
+        }
     }
 
     /// Validate and handle a sum, update or sum2 message.
-    fn handle_message(&mut self, bytes: &[u8]) -> Result<(), PetError> {
-        match self.phase() {
-            Phase::Idle => Err(PetError::InvalidMessage),
-            Phase::Sum => self.handle_sum_message(bytes),
-            Phase::Update => self.handle_update_message(bytes),
-            Phase::Sum2 => self.handle_sum2_message(bytes),
+    pub fn handle_message(&mut self, bytes: &[u8]) -> Result<(), PetError> {
+        let message = self
+            .message_open()
+            .open(&bytes)
+            .map_err(|_| PetError::InvalidMessage)?;
+        let participant_pk = message.header.participant_pk;
+        match (self.phase, message.payload) {
+            (Phase::Sum, PayloadOwned::Sum(msg)) => self.handle_sum_message(participant_pk, msg),
+            (Phase::Update, PayloadOwned::Update(msg)) => {
+                self.handle_update_message(participant_pk, msg)
+            }
+            (Phase::Sum2, PayloadOwned::Sum2(msg)) => self.handle_sum2_message(participant_pk, msg),
+            _ => Err(PetError::InvalidMessage),
         }
     }
 
     /// Validate and handle a sum message.
-    fn handle_sum_message(&mut self, bytes: &[u8]) -> Result<(), PetError> {
-        let msg = SumMessage::open(bytes, self.pk(), self.sk())?;
-        msg.certificate().validate()?;
-        self.validate_sum_task(msg.sum_signature(), msg.pk())?;
-        self.add_sum_participant(msg.pk(), msg.ephm_pk())?;
+    fn handle_sum_message(
+        &mut self,
+        pk: ParticipantPublicKey,
+        message: SumOwned,
+    ) -> Result<(), PetError> {
+        self.validate_sum_task(&pk, &message.sum_signature)?;
+        self.sum_dict.insert(pk, message.ephm_pk);
```

Comment:
this omits the check previously performed in `add_sum_participant()` if the key is already present. therefore replaying messages or sending new valid messages with different contents will overwrite the dict entry now. we can probably leave it like this for now, but we need to define the behavior of the coordinator for repeated messages or for valid messages with different content from the same participant in the future, basically ignore or reject or overwrite.

same holds for the update and sum2 messages, especially since the behavior of sum and update messages on one hand and sum2 messages on the other hand is not consistent anymore.

Reply:
i'd say we postpone this question to another mr.